File: terminal_macros.js
// terminal_macros.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');
const { spawn, execSync } = require('child_process');

const MACROS_FILE = path.join(os.homedir(), '.terminal_macros.json');

class TerminalMacros {
    constructor() {
        this.macros = {};
        this.recording = null;
        this.recordedCmds = [];
        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            prompt: '> '
        });
        this.loadMacros();
    }

    loadMacros() {
        try {
            const data = fs.readFileSync(MACROS_FILE, 'utf8');
            this.macros = JSON.parse(data);
        } catch (err) {
            this.macros = {};
        }
    }

    saveMacros() {
        fs.writeFileSync(MACROS_FILE, JSON.stringify(this.macros, null, 2));
    }

    runShell(cmd) {
        // Используем оболочку
        const shell = process.env.SHELL || (process.platform === 'win32' ? 'cmd' : '/bin/sh');
        const args = process.platform === 'win32' ? ['/c', cmd] : ['-c', cmd];
        const child = spawn(shell, args, { stdio: 'inherit' });
        return new Promise((resolve) => {
            child.on('close', resolve);
        });
    }

    async handleCommand(line) {
        line = line.trim();
        if (!line) return true;

        const parts = line.split(/\s+/);
        const cmd = parts[0];
        const args = parts.slice(1);

        switch (cmd) {
            case 'record':
                if (args.length === 0) {
                    console.log('Usage: record <name>');
                    return true;
                }
                const name = args[0];
                if (this.recording) {
                    console.log(`Already recording macro '${this.recording}'. Use 'stop' to finish.`);
                    return true;
                }
                this.recording = name;
                this.recordedCmds = [];
                console.log(`Recording macro '${name}'. Type 'stop' to finish.`);
                return true;

            case 'stop':
                if (!this.recording) {
                    console.log('No macro is currently being recorded.');
                    return true;
                }
                if (this.recordedCmds.length > 0) {
                    this.macros[this.recording] = this.recordedCmds;
                    this.saveMacros();
                    console.log(`Macro '${this.recording}' recorded (${this.recordedCmds.length} commands).`);
                } else {
                    console.log(`Macro '${this.recording}' is empty, not saved.`);
                }
                this.recording = null;
                this.recordedCmds = [];
                return true;

            case 'play':
                if (args.length === 0) {
                    console.log('Usage: play <name>');
                    return true;
                }
                const macroName = args[0];
                const cmds = this.macros[macroName];
                if (!cmds) {
                    console.log(`Macro '${macroName}' not found.`);
                    return true;
                }
                console.log(`Executing macro '${macroName}'...`);
                for (const c of cmds) {
                    console.log(`> ${c}`);
                    await this.runShell(c);
                }
                return true;

            case 'list':
                if (Object.keys(this.macros).length === 0) {
                    console.log('No macros defined.');
                } else {
                    for (const [name, cmds] of Object.entries(this.macros)) {
                        console.log(`${name} (${cmds.length} commands)`);
                    }
                }
                return true;

            case 'delete':
                if (args.length === 0) {
                    console.log('Usage: delete <name>');
                    return true;
                }
                const delName = args[0];
                if (this.macros[delName]) {
                    delete this.macros[delName];
                    this.saveMacros();
                    console.log(`Macro '${delName}' deleted.`);
                } else {
                    console.log(`Macro '${delName}' not found.`);
                }
                return true;

            case 'help':
                console.log('Commands:');
                console.log('  record <name>   Start recording a macro');
                console.log('  stop            Stop recording');
                console.log('  play <name>     Execute a macro');
                console.log('  list            Show all macros');
                console.log('  delete <name>   Delete a macro');
                console.log('  help            Show this help');
                console.log('  exit            Exit the program');
                console.log('Any other command is executed in the shell.');
                return true;

            case 'exit':
                return false;

            default:
                if (this.recording) {
                    this.recordedCmds.push(line);
                    console.log(`[recording] ${line}`);
                }
                await this.runShell(line);
                return true;
        }
    }

    async run() {
        console.log('Terminal Macros v1.0. Type "help" for commands.');
        this.rl.prompt();
        for await (const line of this.rl) {
            const cont = await this.handleCommand(line);
            if (!cont) break;
            this.rl.prompt();
        }
        this.saveMacros();
        process.exit(0);
    }
}

const app = new TerminalMacros();
app.run().catch(console.error);

File: terminal_macros.py
# terminal_macros.py
import os
import sys
import json
import subprocess
import readline
import atexit
from pathlib import Path

# Файл для хранения макросов
MACROS_FILE = os.path.expanduser("~/.terminal_macros.json")

class TerminalMacros:
    def __init__(self):
        self.macros = {}
        self.recording = None      # имя макроса, если идёт запись
        self.recorded_commands = []
        self.load_macros()

    def load_macros(self):
        if os.path.exists(MACROS_FILE):
            try:
                with open(MACROS_FILE, 'r') as f:
                    self.macros = json.load(f)
            except:
                self.macros = {}
        else:
            self.macros = {}

    def save_macros(self):
        with open(MACROS_FILE, 'w') as f:
            json.dump(self.macros, f, indent=2)

    def run_shell_command(self, cmd):
        """Выполняет команду в оболочке и выводит результат."""
        try:
            result = subprocess.run(cmd, shell=True, text=True, capture_output=True)
            if result.stdout:
                print(result.stdout, end='')
            if result.stderr:
                print(result.stderr, end='', file=sys.stderr)
            return result.returncode
        except Exception as e:
            print(f"Error: {e}", file=sys.stderr)
            return 1

    def handle_command(self, line):
        line = line.strip()
        if not line:
            return True

        # Обработка встроенных команд
        if line.startswith('record '):
            name = line[7:].strip()
            if not name:
                print("Usage: record <name>")
                return True
            if self.recording:
                print(f"Already recording macro '{self.recording}'. Use 'stop' to finish.")
                return True
            self.recording = name
            self.recorded_commands = []
            print(f"Recording macro '{name}'. Type 'stop' to finish.")
            return True

        if line == 'stop':
            if not self.recording:
                print("No macro is currently being recorded.")
                return True
            if self.recorded_commands:
                self.macros[self.recording] = self.recorded_commands
                self.save_macros()
                print(f"Macro '{self.recording}' recorded ({len(self.recorded_commands)} commands).")
            else:
                print(f"Macro '{self.recording}' is empty, not saved.")
            self.recording = None
            self.recorded_commands = []
            return True

        if line.startswith('play '):
            name = line[5:].strip()
            if not name:
                print("Usage: play <name>")
                return True
            if name not in self.macros:
                print(f"Macro '{name}' not found.")
                return True
            print(f"Executing macro '{name}'...")
            for cmd in self.macros[name]:
                print(f"> {cmd}")
                self.run_shell_command(cmd)
            return True

        if line == 'list':
            if not self.macros:
                print("No macros defined.")
            else:
                for name, cmds in self.macros.items():
                    print(f"{name} ({len(cmds)} commands)")
            return True

        if line.startswith('delete '):
            name = line[7:].strip()
            if not name:
                print("Usage: delete <name>")
                return True
            if name in self.macros:
                del self.macros[name]
                self.save_macros()
                print(f"Macro '{name}' deleted.")
            else:
                print(f"Macro '{name}' not found.")
            return True

        if line == 'help':
            print("Commands:")
            print("  record <name>   Start recording a macro")
            print("  stop            Stop recording")
            print("  play <name>     Execute a macro")
            print("  list            Show all macros")
            print("  delete <name>   Delete a macro")
            print("  help            Show this help")
            print("  exit            Exit the program")
            print("Any other command is executed in the shell.")
            return True

        if line == 'exit':
            return False

        # Если идёт запись, добавляем команду в макрос
        if self.recording is not None:
            self.recorded_commands.append(line)
            print(f"[recording] {line}")

        # Выполняем команду в оболочке
        self.run_shell_command(line)
        return True

    def run(self):
        print("Terminal Macros v1.0. Type 'help' for commands.")
        while True:
            try:
                line = input("> ")
                if not self.handle_command(line):
                    break
            except KeyboardInterrupt:
                print("\nUse 'exit' to quit.")
            except EOFError:
                break
        self.save_macros()

if __name__ == "__main__":
    app = TerminalMacros()
    app.run()
